README.md
# Flask-RQ

.. image:: https://travis-ci.org/mattupstate/flask-rq.svg?branch=master
    :target: https://travis-ci.org/mattupstate/flask-rq

RQ (Redis Queue) integration for Flask applications

## Pallets Community Ecosystem

> [!IMPORTANT]\
> This project is part of the Pallets Community Ecosystem. Pallets is the open
> source organization that maintains Flask; Pallets-Eco enables community
> maintenance of related projects. If you are interested in helping maintain
> this project, please reach out on [the Pallets Discord server][discord].

[discord]: https://discord.gg/pallets

## Resources

- `Documentation <http://packages.python.org/Flask-RQ/>`_
- `Issue Tracker <http://github.com/mattupstate/flask-rq/issues>`_
- `Code <http://github.com/mattupstate/flask-rq/>`_
- `Development Version
  <http://github.com/mattupstate/flask-rq/zipball/develop#egg=Flask-RQ-dev>`_


## Installation

```bash

    $ pip install flask-rq
```

## Getting started

To quickly start using `rq`, simply create an RQ instance:

```python

    from flask import Flask
    from flask.ext.rq import RQ


    app = Flask(__name__)

    RQ(app)
```

### ``@job`` decorator

Provides a way to quickly set a function as an ``rq`` job:

```python

    from flask.ext.rq import job


    @job
    def process(i):
        #  Long stuff to process


    process.delay(3)
```

A specific queue name can also be passed as argument:

```python

    @job('low')
    def process(i):
        #  Long stuff to process


    process.delay(2)
```

### ``get_queue`` function

Returns default queue or specific queue for name given as argument:

```python

    from flask.ext.rq import get_queue


    job = get_queue().enqueue(stuff)  # Creates a job on ``default`` queue
    job = get_queue('low').enqueue(stuff)  # Creates a job on ``low`` queue
```

### ``get_worker`` function

Returns a worker for default queue or specific queues for names given as arguments:

```python
    from flask.ext.rq import get_worker


    # Creates a worker that handle jobs in ``default`` queue.
    get_worker().work(True)
    # Creates a worker that handle jobs in both ``default`` and ``low`` queues.
    get_worker('default', 'low').work(True)
    # Note: These queues have to share the same connection
```

## Configuration

By default Flask-RQ will connect to the default, locally running
Redis server. One can change the connection settings for the default
server like so:

```python
    app.config['RQ_DEFAULT_HOST'] = 'somewhere.com'
    app.config['RQ_DEFAULT_PORT'] = 6479
    app.config['RQ_DEFAULT_PASSWORD'] = 'password'
    app.config['RQ_DEFAULT_DB'] = 1
```

Queue connection can also be set using a DSN:

```python

    app.config['RQ_LOW_URL'] = 'redis://localhost:6379/2'
```



CHANGES
Flask-RQ Changelog
==================

Here you can see the full list of changes between each Flask-RQ release.

Version 0.3
-----------

Released August 14 2013

- Added `ctx_delay` method. Allow to work inside of flask context in delayed job


Version 0.2
-----------

Released November 12th 2012

- Improve and simplify extension
- Renamed `task` to `job`


Version 0.1
-----------

First release

flask_rq.py
# -*- coding: utf-8 -*-
"""
    flask_rq
    ~~~~~~~~

    RQ (Redis Queue) integration for Flask applications.

    :copyright: (c) 2012 by Matt Wright.
    :license: MIT, see LICENSE for more details.

"""

__version__ = '0.2'

import redis

from flask import current_app
from redis._compat import urlparse
from rq import Queue, Worker

_flask_app = None

default_config = {
    'RQ_DEFAULT_HOST': 'localhost',
    'RQ_DEFAULT_PORT': 6379,
    'RQ_DEFAULT_PASSWORD': None,
    'RQ_DEFAULT_DB': 0
}


def config_value(name, key):
    name = name.upper()
    config_key = 'RQ_%s_%s' % (name, key)
    app = _flask_app if _flask_app else current_app
    if not config_key in app.config \
            and not 'RQ_%s_URL' % name in app.config:
        config_key = 'RQ_DEFAULT_%s' % key
    return app.config.get(config_key, None)


def get_connection(queue='default'):
    url = config_value(queue, 'URL')
    if url:
        return redis.from_url(url, db=config_value(queue, 'DB'))
    return redis.Redis(host=config_value(queue, 'HOST'),
        port=config_value(queue, 'PORT'),
        password=config_value(queue, 'PASSWORD'),
        db=config_value(queue, 'DB'))


def get_queue(name='default', **kwargs):
    kwargs['connection'] = get_connection(name)
    return Queue(name, **kwargs)


def get_server_url(name):
    url = config_value(name, 'URL')
    if url:
        url_kwargs = urlparse(url)
        return '%s://%s' % (url_kwargs.scheme, url_kwargs.netloc)
    else:
        host = config_value(name, 'HOST')
        password = config_value(name, 'HOST')
        netloc = host if not password else ':%s@%s' % (password, host)
        return 'redis://%s' % netloc


def get_worker(*queues):
    if len(queues) == 0:
        queues = ['default']
    servers = [get_server_url(name) for name in queues]
    if not servers.count(servers[0]) == len(servers):
        raise Exception('A worker only accept one connection')
    return FlaskRQWorker([get_queue(name) for name in queues],
        connection=get_connection(queues[0]))


def job(func_or_queue=None):
    if callable(func_or_queue):
        func = func_or_queue
        queue = 'default'
    else:
        func = None
        queue = func_or_queue

    def wrapper(fn):
        def delay(*args, **kwargs):
            q = get_queue(queue)
            return q.enqueue(fn, *args, **kwargs)

        fn.delay = delay
        return fn

    if func is not None:
        return wrapper(func)

    return wrapper


class RQ(object):
    def __init__(self, app=None):
        if app is not None:
            self.init_app(app)

    def init_app(self, app):
        global _flask_app
        for key, value in default_config.items():
            app.config.setdefault(key, value)
        _flask_app = app


class FlaskRQWorker(Worker):
    def __init__(self, *args, **kwargs):
        return super(FlaskRQWorker, self).__init__(*args, **kwargs)

    def perform_job(self, *args, **kwargs):
        with _flask_app.app_context():
            return super(FlaskRQWorker, self).perform_job(*args, **kwargs)
